src/NHibernate/Driver/OracleManagedDataClientDriver.cs

```diff
@@ -47,8 +48,31 @@ public OracleManagedDataClientDriver()
 			oracleDbTypeRefCursor = Enum.Parse(oracleDbTypeEnum, "RefCursor");
 			oracleDbTypeXmlType = Enum.Parse(oracleDbTypeEnum, "XmlType");
 			oracleDbTypeBlob = Enum.Parse(oracleDbTypeEnum, "Blob");
+			_oracleDbTypeNVarchar2 = Enum.Parse(oracleDbTypeEnum, "NVarchar2");
 		}
 
+		/// <summary>
+		/// By default, ODP.Net maps <see cref="DbType.String"/> to <c>OracleDbType.Varchar2</c>. If you need it to be mapped
+		/// to <c>OracleDbType.NVarchar2</c> instead, set this property to <see langword="true"/>.
+		/// </summary>
+		/// <remarks>
+		/// <para>
+		/// This is not a bug. Varchar2 with Oracle can be Unicode, depending on the server configuration. If the server is
+		/// configured for having it Unicode, Varchar2 should be used for Unicode string. NVarchar2 is there for allowing
+		/// storing Unicode string on server not configured for Unicode.
+		/// http://docs.oracle.com/cd/E51173_01/win.122/e17732/featOraCommand.htm#i1007432
+		/// </para>
+		/// <para>
+		/// So the default NHibernate choice in Oracle dialects to map <see cref="DbType.String"/> to <c>nvarchar2</c> is a bit
+		/// unfortunate, because this is not the default mapping for Oracle. For tests, since the database is created by NHibernate,
+		/// string columns get typed <c>nvarchar2</c>, but parameters are still typed <c>varchar2</c>, causing an
+		/// <c>ORA-12704: character set mismatch</c> for some queries. This being not even reported, I guess most NHibernate Oracle
+		/// users handles their tables themselves and types them with <c>varchar2</c>, avoiding the trouble. So changing the behavior
+		/// for parameters type would be a major breaking change.
+		/// </para>
+		/// </remarks>
+		public bool UseNVarchar2ForStringParameter { get; set; }
```

Comment:
This is the wrong assumption, I belive.

We have 4 different mappings of strings in ADO.NET:
1. AnsiString - VARCHAR(n) (VARCHAR2(n) in the case of Oracle) - the single byte string collation dependent.
2. AnsiStringFixedLength - CHAR(n) - fixed length string.
3. String - NVARCHAR(n) (NVARCHAR2(n) in the case of Oracle) - always Unicode strings (UTF16, I think)
4. StringFixedLength - NCHAR(n) - fixed length Unicode string (UTF16, I think).

We shall really fix the provider to obey this rule.